### PR TITLE
raise BG target for noisy / raw CGM data

### PR DESCRIPTION
**Pull request by scottleibrand**
When looping off of noisy / raw CGM data (during sensor swaps, on noisy sensors, etc.) it would be safest to adjust the BG target upward to provide an extra safety buffer for erroneous or miscalibrated CGM data.  This PR defaults the noisyCGMTargetMultiplier to 1.3, and adjusts the default maxraw value accordingly, from 150 to 200.  This will allow looping off raw/noisy CGM data values up to 200 mg/dL, but with an elevated target of 130% of normal.

This incorporates #735 to do the same for xDrip noise values >= 2 as for Dexcom raw BG values <= 200.  Dexcom raw values > 200, and xDrip noise values >= 3, will prevent looping entirely.

There is one remaining TODO, to support maxRaw in preferences.json.  The g4-upload config current does everything in a separate myopenaps-cgm-loop directory, so can't read myopenaps/preferences.json.  We'll probably want to refactor g4-upload to be more like g4-upload, which will simplify the config a lot and also make it much easier to read preferences.json.

Thoughts?

### CODE REVIEW

**Comment by alimhassam** [comment]
Reviewed with the latest changes, the noise approach seems sound to me (except the small issue might have been due to merging which needs to be corrected).

For the raw glucose logic, I don't think i fully understand it. Not fully related to this pr, but it seems like it's trying to use calibration records from the old sensor to calculate bg for the new sensor on sensor changes? Did I get this right? Is that a safe thing to do?

**Comment by scottleibrand** [comment]
@alimhassam I don't see your review noting the small issue.  Did you hit Submit on it?

We do not explicitly recommend looping on an uncalibrated sensor (or a receiver calibrated from a previous sensor), but have personally seen better outcomes doing that (with appropriate limitations on maxRaw and with appropriate BG or target adjustments, as are in this PR) rather than stopping looping entirely for 2h.  More commonly, those using Dexcom receivers will restart the receiver's sensor session after 7 days without actually replacing the sensor, in which case the old calibration record is still perfectly good for the new session.

File: lib/determine-basal/determine-basal.js [hunk 1]
```diff
@@ -65,8 +65,11 @@ var determine_basal = function determine_basal(glucose_status, currenttemp, iob_
     var minAgo = round( (systemTime - bgTime) / 60 / 1000 ,1);
 
     var bg = glucose_status.glucose;
-    if (bg < 39) {  //Dexcom is in ??? mode or calibrating
-        rT.reason = "CGM is calibrating or in ??? state";
+    var noise = glucose_status.noise;
+    // 38 is an xDrip error state that usually indicates sensor failure
+    // all other BG values between 11 and 37 mg/dL reflect non-error-code BG values, so we should zero temp for those
+    if (bg <= 10 || bg == 38 || noise >= 3) {  //Dexcom is in ??? mode or calibrating, or xDrip reports high noise
```
[the remark this answers is not in the document]
Comment (alimhassam) [reply lib/determine-basal/determine-basal.js]: oops, didn't realise i needed to press submit.

[the remark this answers is not in the document]
Comment (scottleibrand) [reply lib/determine-basal/determine-basal.js]: Yes, good catch, thanks.

**Comment by alimhassam** [comment]
Right, makes sense for the old sensor case, but i'm a bit skeptical/surprised it's safe on a new sensor.

**Comment by tim2000s** [comment]
Let me try out the noise management approach that Adrian mentioned in the issue before merging this in. 

Ideally I’d like to be able to choose whether it goes more conservative at 2 or 3 and the behaviour based on some longer term testing before merging.

**Comment by scottleibrand** [comment]
Ok, I’ll wait a bit. We can always make further adjustments after it’s in dev, though. Once we’re confident this is better than the status quo we can go ahead and merge it for further testing in dev.

**Comment by drnoname82** [comment]
We have tested this PR with XDrip+ and it works as intended. I think this is a great safety improvement!

File: lib/determine-basal/determine-basal.js [hunk 2]
```diff
@@ -272,8 +275,21 @@ var determine_basal = function determine_basal(glucose_status, currenttemp, iob_
     // adjust that for deviation like we did eventualBG
     //var snoozeBG = naive_snoozeBG + deviation;
 
-    // adjust target BG range if needed to safely bring down high BG faster without causing lows
-    if ( bg > max_bg && profile.adv_target_adjustments && ! profile.temptargetSet ) {
+    // raise target for noisy / raw CGM data
+    if (glucose_status.noise >= 2) {
```
Comment (alimhassam) [inline lib/determine-basal/determine-basal.js]: Should we say something in the reason field when this situation occurs?

**Comment by alimhassam** [comment]
Been running a couple of days with this change as well and I works well for me so far on xdrip+ and g5. 


Generally very happy with the change, definitely an improvement.

**Comment by drnoname82** [comment]
We just had a sensor failure (XDrip+/G5) and everything worked as intended. I just saw in @jamorham 's comment https://github.com/NightscoutFoundation/xDrip/issues/191#issuecomment-334175061 that noise can also be "0" which is an error state or means that noise cannot be computed. I think that we should thus also exclude BG values where noise=0.  (I have zero readings in our database where this is the case, so this will not prevent looping on useful values)

**Comment by scottleibrand** [comment]
Based on the feedback from @drnoname82 and @alimhassam I think we should go ahead and merge this and make any further tweaks in a new PR.

**Comment by drnoname82** [comment]
Thumbs up! The PR has also proven quite useful in avoiding useless changes in insulin delivery due to compression lows and subsequent sharp rises, usually accompanied by high noise.

**Comment by tim2000s** [comment]
I’ve no issue with this pr stopping looping at noise level 3, but at noise level 2 I don’t see enough variation from 1 for it to cause an issue with looping. 

I think the level of noise a user accepts should be a hidden preference. Especially when you consider how many people are extending the life of a sensor. 

For me, I’d prefer to continue to loop at 2 and remove SMB at 3. I’m comfortable that the smoothing components of the original oref0 algorithm manage unexpected deviations effectively when using a TBR-only model.

**Comment by alimhassam** [comment]
@tim2000s this changes is done in such a way that looping continues at noise level 2, but slightly more conservatively, was that your understanding? 

(I've been using thing for more than 1 week, and it works well, for me)

**Comment by tim2000s** [comment]
Yes, that is my understanding. My point is that I’ve been looping with xdripaps for the best part of a year. In that time the periods where Dex reported 3 (xDrip didn’t for a long time) were problematic when using SMB but the look back in the oref0 TBR based code actually allowed it to loop effectively with highly noisy readings. 

The noise level 2 in my observations seems really rare and to have a very low deviation from 1. I don’t think we should just PR something into Dev because it “feels” safer without properly quantifying the improvement. From a purely observational point of view, the 3 level of noise really needs to shut down SMB and I’d agree that it’s marginal that some would continue to loop and others wouldn’t using TBRs. 

At noise level 2 I don’t see enough deviation nor a long enough duration of 2s to justify making the targets more conservative. I’m happy to be quantifiably proved wrong, but that’s my point of view.

**Comment by alimhassam** [comment]
Ok, i won't argue your suggested behavior for noise level 3, there might be way to still loop safely, and improvements around this might be interesting.

For noise level 2, if they rarely occur in your case, than the behavior here won't affect you much (or rarely). 

I personally get noise=2 almost daily. I did finger pricks comparison a couple of times, and compared with the rest of the curve around the time, at it generally reflects a jump in the unfiltered reading, which granted, gets somwhat smoothed out by the xdrip+ algorithm, but does represents uncertainty. 

I think being more conservative around such uncertainty a good thing. Before merging this branch I did get one occurrence where smb activated for a mild bg jump with noise=2. There were Cob on, which resulted in an extra bolus and a followup low, and that's the reason i started looking at noise issues more closely.

Of course, if you have access to more data, and are able to get "proper quantifying" that would be much better. 
There is already a preference setting to change how much you want to adjust the target for noisy readings. I think the default adjustment is relatively mild at 1.3, but if you If you want update the pr to improve and make this setting more granular or do something different, or add other ones maybe for noise=3 that might be good too, as long as this current behavior can still be achieved.

Personally (with my limited experience&data) the current implemented behavior seems to fix the issue i was having with noisy readings.

**Comment by drnoname82** [comment]
I just went through the past few days of CGM data. My original opinion was to only loop off CGM data with a noise level of 1. However, I have to agree with @tim2000s that at least our data does not justify that. We see ca. 5% data with noise=2. I failled to find any obviously wrong readings within the past week. 

How about merging this PR as is and gathering more data? It definitely is a safety improvement and I find looping with a raised target for ca. 5% of the time completely acceptable. If others see much more noise=2 readings, that might be a different situation. I think most users are not even aware that XDrip now provides noise levels and we will only get more opinions if this is merged to dev.

**Comment by tim2000s** [comment]
It’s tricky because the inherent functionality built into oref0 is halfway to a Kaplan filter, limiting the variation that the future predictions will have by looking at past deviations and comparing them to what the expected behaviour is, rather than relying on the individual readings. 

My preferred (but much more complex solution) would be to apply a Kaplan filter based on the noise data rather than simply shut off looping or increase the target. That does require work though.

**Comment by alimhassam** [comment]
That's an interesting idea @tim2000s you basically want to update the oref0 algorithm to work of the raw filtered and unfiltered data and do its own noise working. This could get better error estimate rather that just a single noise number, as that would be part of the kalman filter output.

xdrip+ already implements such noise workings but only outputs a single number rather that full error variance.

But I suspect that once the error variance estimate is produced, thr resulting dosing decision will end up being very similar to what was done in this pr. Ie raise BG target when error estimate increase. 
The advantage of your approach is that since the error might be more quantifiable and we can change the target more precisely than only based on the less precise noise number.

But I think the main idea would be the same. Correct me if I missed or misunderstood something.

**Comment by drnoname82** [comment]
We have been testing this for weeks without any issues. Yesterday we had a situation where a perfectly fine sensor drifted too low and then jumped significantly too high (CGM 270, BG 190) with noise reported as "light" for about twenty minutes, afterwards noise was reported as high/extreme. So I think treating CGM readings with light noise conservatively is more than justified.

The reasoning in Nightscout is clear for noise values >= 3 but I did never see any reasoning in the OpenAPS pill when target was raised due to noise=2.

I only just saw that there is a preference `NoisyCGMTargetMultiplier` allowing anyone wishing to do so trust those CGM values more. I think this PR is good to merge since by providing this variable it mean s hardly any limitation for those wishing to loop off lightly noisy values.

**Comment by scottleibrand** [comment]
Hearing no objections that aren’t addressed by the ability to set preferences, I’m merging this to dev.

**Pull request merged**